[Core] New feature : User Interface to "Detect" Feed from an URL

sysadminstory opened the pull request:
Detect Action has been expanded to support returning a Feed in a JSON
format instead of a Redirect. Existing usage of the Detect action will
keep working as usual.

Frontpage template has now a section to display the Feed detection
result, and a button to start the Feed Detection.

An information is displayed if a Feed is found or not, to let know that even if there are no Feed found, or one Feed found, there may be other Feed available using the classical Feed creation.

A new JS file contains the necessary JS (Ajax and Event management) to
fill the Feed Detection section.

Feel Free to comment and suggest changes !

sysadminstory committed:
[Core] New feature : User Interface to "Detect" Feed from an URL

Detect Action has been expanded to support returning a Feed in a JSON
format instead of a Redirect. Existing usage of the Detect action will
keep working as usual.

Frontpage template has now a section to display the Feed detection
result, and a button to start the Feed Detection.

A new JS file contains the necessary JS (Ajax and Event management) to
fill the Feed Detection section.

sysadminstory committed:
Coding policy fixes

dvikan commented:
1. just so you know, the existing search bar searches for url too as defined in the bridge

2. use the modern `fetch()` instead of oldschool ajax

3. i like the feature but looks kinda ugly on the front page. it's a ux challenge

4. possibly create a completely new *action* instead of entangling it with the `Detect` action.

5. there is template literals which are ES6 if we are brave: https://caniuse.com/?search=template%20literal

6. ideally the detect feature and the new feature in this pr should support returning multiple results

7. this detect feature should return an absolute url starting with `https://` but i understand the prev behaviour also not do that

8.  the frontend code could be improved. will become difficult to maintain if it grows

9. check out `Json::encode()`

@Mynacol @Bockiii

Bockiii commented:
I have no opinion on this tbh. I dont really understand the use case or which type of user would benefit from this, so ... no opinion

sysadminstory commented:
> I have no opinion on this tbh. I dont really understand the use case or which type of user would benefit from this, so ... no opinion

The not techy user of a public instance could get a feed from an existing URL, without having to enter the parameters manually in the Bridge card.

Mynacol commented on static/detectbridge.js:
```diff
@@ -0,0 +1,70 @@
+function displayDetectionFail(data)
+{
+    const detectresults = document.getElementById('detectresults');
+    detectresults.innerHTML = 'No Feed found !<div class="alert alert-info" role="alert">Every Bridge does not support feed detection, you can check within the bridge below to create a Feed using the bridge parameters</div>';
```
Spelling.

Mynacol commented:
I like the idea, just pasting a website link and getting a rss-bridge link would be nice.

I tried to use it with a YouTube channel link but it didn't create any feed. Is this only the groundwork? Do bridges have to add a detection method? How would such a method look like? Some documentation should be added then.

I agree with the review of dvikan, and want to add some more points:

1. If you stick to using the existing DetectAction (which I slightly oppose), you should at least reuse the existing `format` URI parameter, instead of creating a new `output` parameter. Having `...&format=Html&output=json&...` gets confusing real quick.

This PR currently requires significant rewrites. I would be glad if you tackle the criticized points :)

sysadminstory commented:
> 1. just so you know, the existing search bar searches for url too as defined in the bridge

I know it filters on the URL doing the write :)
> 
>     2. use the modern `fetch()` instead of oldschool ajax

I'm not a big fan of Javascript Promise, but I will use fetch instead of Ajax !

 
>     3. i like the feature but looks kinda ugly on the front page. it's a ux challenge

That's the purpose of this PR :D
Would it be better with a separate page ? or a separate section on the front page ?
How would you enhance the existing UX ?
 
>     4. possibly create a completely new _action_ instead of entangling it with the `Detect` action.

Ok !

>     5. there is template literals which are ES6 if we are brave: https://caniuse.com/?search=template%20literal

If I stick to a "Dynamic  page" approach, I will use template literals to get a more readable code !

>     6. ideally the detect feature and the new feature in this pr should support returning multiple results

I will create a new action, so I will allow to return multiple results !

>     7. this detect feature should return an absolute url starting with `https://` but i understand the prev behaviour also not do that

OK, I will change to absolute URL in the new action !

>     8. the frontend code could be improved. will become difficult to maintain if it grows

Can you be more precise ?

>     9. check out `Json::encode()`

OK ! I was not aware of this class !

sysadminstory commented:
> I like the idea, just pasting a website link and getting a rss-bridge link would be nice.

Me too :D

> I tried to use it with a YouTube channel link but it didn't create any feed. Is this only the groundwork? Do bridges have to add a detection method? How would such a method look like? Some documentation should be added then.

I'm using the existing feature of some bridges. They implement the function `detectParameters` that is already documented :
https://github.com/RSS-Bridge/rss-bridge/blob/master/docs/05_Bridge_API/02_BridgeAbstract.md#detectparameters

You can test it for example with the TikTok bridge :)

> I agree with the review of dvikan, and want to add some more points:
> 
>     1. If you stick to using the existing DetectAction (which I slightly oppose), you should at least reuse the existing `format` URI parameter, instead of creating a new `output` parameter. Having `...&format=Html&output=json&...` gets confusing real quick.
> 
> 
> This PR currently requires significant rewrites. I would be glad if you tackle the criticized points :)

I will use a different action, to not complicate DetectAction !

sysadminstory committed:
[Core] New feature : User Interface to "Detect" Feed from an URL

- Switch from old school XMLHttpRequest to fetch
- Enhance UX of search results
- Revert to it's original content
- Switch to a new Action : FindfeedAction.php
- Switch to template literals instead of string concatenation
- FindFeed action could retrun multiple feeds
- Results are sent with an absolute URL
- Switch to Json::encode() helper function

sysadminstory commented:
So, I have followed almost all recommendation !

The only recommendation I did not understand was :
> 8. the frontend code could be improved. will become difficult to maintain if it grows

Can you help me to understand what you mean here ?
@dvikan

sysadminstory committed:
[Core] New feature : User Interface to "Detect" Feed from an URL

- Move specific JS code to rss-bridge.js
- Change HTML tag for the button to have a consistant style with th rest
  of the page

sysadminstory committed:
[Core] New feature : User Interface to "Detect" Feed from an URL

- If no context is sent, assume there is only one unnamed context
- Find parameter name in global and currect context

sysadminstory commented:
I've standardized the interface design, and taken into account the case of global parameters and unique/unnamed contexts

dvikan committed:
fix

dvikan committed:
remove typo

dvikan commented:
i dont think the async keyword is useful here. im a bit oldschool though havent caught up with the crazies

shits ugly but lets try it out. can always remove if the shit hits the fan
